File: internal/agentrun/event_test.go
package agentrun

import (
	"encoding/json"
	"reflect"
	"strings"
	"testing"
	"time"
)

func TestEventKindContractIsClosed(t *testing.T) {
	kinds := []EventKind{
		EventLifecycle,
		EventTool,
		EventCommand,
		EventFile,
		EventTest,
		EventCommit,
		EventError,
		EventSummary,
	}
	want := []string{
		"lifecycle", "tool", "command", "file",
		"test", "commit", "error", "summary",
	}
	if len(kinds) != len(want) {
		t.Fatalf("kind count = %d, want %d", len(kinds), len(want))
	}
	for i, kind := range kinds {
		if !kind.Valid() || string(kind) != want[i] {
			t.Fatalf("kind[%d] = %q valid=%v, want %q", i, kind, kind.Valid(), want[i])
		}
	}
	for _, invalid := range []EventKind{"", "prompt", "transcript", "reasoning", "future"} {
		if invalid.Valid() {
			t.Fatalf("unexpected valid kind %q", invalid)
		}
	}
}

func TestEventPhaseAndOutcomeContractsAreClosed(t *testing.T) {
	phases := []EventPhase{
		EventStarted,
		EventProgress,
		EventWaiting,
		EventBlocked,
		EventCompleted,
		EventFailed,
	}
	for _, phase := range phases {
		if !phase.Valid() {
			t.Fatalf("phase %q is invalid", phase)
		}
	}
	for _, invalid := range []EventPhase{"", "idle", "done", "thinking"} {
		if invalid.Valid() {
			t.Fatalf("unexpected valid phase %q", invalid)
		}
	}
	for _, outcome := range []EventOutcome{EventSucceeded, EventUnsuccessful} {
		if !outcome.Valid() {
			t.Fatalf("outcome %q is invalid", outcome)
		}
	}
	for _, invalid := range []EventOutcome{"", "unknown", "partial"} {
		if invalid.Valid() {
			t.Fatalf("unexpected valid outcome %q", invalid)
		}
	}
}

func TestEventJSONRoundTripPreservesStructuredEvidence(t *testing.T) {
	exitCode := 1
	now := time.Date(2026, time.August, 10, 18, 0, 0, 0, time.UTC)
	want := Event{
		ModelVersion:   EventModelVersion,
		ID:             "host-event-1",
		RunID:          "run-1",
		Provider:       "codex",
		SourceID:       "provider-event-9",
		SourceSequence: 9,
		HostSequence:   4,
		Kind:           EventTest,
		Phase:          EventFailed,
		Outcome:        EventUnsuccessful,
		Subject:        "go-test",
		Paths:          []string{"internal/agentrun/event_test.go"},
		CommitSHA:      "0123456789abcdef0123456789abcdef01234567",
		ExitCode:       &exitCode,
		ErrorClass:     "test_failure",
		Summary:        "One bounded provider summary.",
		OccurredAt:     now.Add(-time.Second),
		ObservedAt:     now,
	}
	encoded, err := json.Marshal(want)
	if err != nil {
		t.Fatal(err)
	}
	var got Event
	if err := json.Unmarshal(encoded, &got); err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("round trip mismatch\n got: %#v\nwant: %#v", got, want)
	}
}

func TestEventContractsContainNoRawContentOrAuthorityFields(t *testing.T) {
	forbidden := []string{
		"prompt", "message", "reasoning", "chainOfThought", "transcript",
		"args", "arguments", "input", "result", "output", "stdout", "stderr",
		"environment", "headers", "body", "token", "credential", "capability",
		"projectRoot", "planId", "taskId", "terminalId", "cwd",
	}
	for _, contract := range []reflect.Type{
		reflect.TypeOf(EventObservation{}),
		reflect.TypeOf(Event{}),
	} {
		for i := 0; i < contract.NumField(); i++ {
			field := contract.Field(i)
			if field.Type.Kind() == reflect.Map {
				t.Fatalf("%s.%s uses an open-ended map", contract.Name(), field.Name)
			}
			name := strings.Split(field.Tag.Get("json"), ",")[0]
			for _, denied := range forbidden {
				if name == denied {
					t.Fatalf("%s exposes forbidden JSON field %q", contract.Name(), name)
				}
			}
		}
	}
}

File: crates/ptrack-agent/src/event_test.rs
use super::{
    EVENT_MODEL_VERSION, Event, EventCorrelation, EventKind, EventNotificationKind,
    EventObservation, EventOutcome, EventPhase, Timestamp,
};

#[test]
fn event_json_round_trip_preserves_only_structured_evidence() {
    let value = Event {
        model_version: EVENT_MODEL_VERSION,
        id: "host-event-1".to_owned(),
        run_id: "run-1".to_owned(),
        provider: "codex".to_owned(),
        source_id: "provider-event-9".to_owned(),
        source_sequence: 9,
        host_sequence: 4,
        lifecycle_revision: 2,
        kind: EventKind::Test,
        phase: EventPhase::Failed,
        outcome: EventOutcome::Failed,
        subject: "cargo-test".to_owned(),
        paths: vec!["src/event.rs".to_owned()],
        commit_sha: "0123456789abcdef".to_owned(),
        exit_code: Some(1),
        error_class: "test_failure".to_owned(),
        summary: "bounded".to_owned(),
        occurred_at: Timestamp::from_unix_nanoseconds(2_000_000_000),
        observed_at: Timestamp::from_unix_nanoseconds(3_000_000_000),
        correlation: EventCorrelation {
            project_root: "/project".to_owned(),
            ..EventCorrelation::default()
        },
        notification: EventNotificationKind::Unset,
    };
    let encoded = serde_json::to_string(&value).unwrap();
    assert_eq!(serde_json::from_str::<Event>(&encoded).unwrap(), value);
    assert_eq!(
        encoded,
        r#"{"modelVersion":1,"id":"host-event-1","runId":"run-1","provider":"codex","sourceId":"provider-event-9","sourceSequence":9,"hostSequence":4,"lifecycleRevision":2,"kind":"test","phase":"failed","outcome":"failed","subject":"cargo-test","paths":["src/event.rs"],"commitSha":"0123456789abcdef","exitCode":1,"errorClass":"test_failure","summary":"bounded","occurredAt":"1970-01-01T00:00:02Z","observedAt":"1970-01-01T00:00:03Z","correlation":{"projectRoot":"/project"}}"#
    );
}

#[test]
fn event_observation_cannot_deserialize_adapter_recognition() {
    let decoded: EventObservation = serde_json::from_str(r#"{"modelVersion":1,"sourceId":"x","sourceSequence":1,"kind":"lifecycle","phase":"waiting","notification":"question","recognizedNotification":true}"#).unwrap();
    assert!(!decoded.recognized_notification);
    assert_eq!(decoded.notification, EventNotificationKind::Question);
}

#[test]
fn event_contract_enums_are_closed() {
    assert!(EventKind::Lifecycle.is_valid());
    assert!(EventPhase::Completed.is_valid());
    assert!(EventOutcome::Succeeded.is_valid());
    assert!(EventNotificationKind::Completion.is_valid());
    assert!(serde_json::from_str::<EventKind>(r#""prompt""#).is_err());
    assert!(serde_json::from_str::<EventPhase>(r#""idle""#).is_err());
}
